src/exchange.c

```diff
@@ -267,17 +267,24 @@ enum wps_result do_wps_exchange()
 		  (last_msg == M3 || last_msg == M5))
 		{
 			ret_val = KEY_REJECTED;
-			/* Got timeout instead of an M5 message, when cracking second half */
-			if (!get_pin_string_mode() && last_msg == M3 && get_key_status() == KEY2_WIP) {
-				ret_val = UNKNOWN_ERROR;
-				cprintf(WARNING, "[!] WARNING: Potential first half pin has changed!\n");
-			}
+		}
+		/*
+		 * Some WPS implementations sending deauth request instead of sending a NACK.
+		 * Treat the timeout as NACK if receive deauth request while waiting for M5/M7.
+		 */
+		else if (deauth_flag && (last_msg == M3 || last_msg == M5) && get_deauth_is_nack()) {
+			ret_val = KEY_REJECTED;
```

Comment:
if we set this to REJECTED, wouldn't it mean that even though the pin could be correct when router sends deauth, next pin will be tried ?

Reply:
ah, sorry that i didn't mention it earlier, but my idea was not to introduce another command line option, but to always Do The Right Thing, in which case this behaviour would need to be fine-tuned

Reply:
With WPS 2.0, it can't determine whether or not router sends deauth request instead of NACK in just a few attempts because the WPS LOCK.
The solution I thought would be to store the count of receiving deauth request without NACK in .wpc file.
If after 10 count, for example, not received any NACK then it can consider router sends deauth request instead of NACK.
The count starts with 0, increments with each deauth request without NACK, if receive NACK then store -1 that represents router sends NACK and ignore deauth request treatment.